### PR TITLE
Make dashboard status widget compatible with HPOS

### LINKED ISSUE #40382: HPOS: WooCommerce status dashboard widget uses wp_count_posts
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When using HPOS `High performance order storage ` as the data store with compatibility mode off, the WooCommerce  status widget does not correctly reflect order data.

This is because it uses `wp_count_posts` [here](https://github.com/woocommerce/woocommerce/blob/55c1bec4900137bb69405ae9d714ed5288ae2939/plugins/woocommerce/includes/admin/class-wc-admin-dashboard.php#L204-L207)

### Expected behavior

The dashboard remains reliable

### Steps to reproduce

1. Enable HPOS
2. Disable maintain sync with posts
3. Place a new order.
4. The order will not reflect in the WooCommerce status dashboard

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): http://blueprint.local
Site address (URL): http://blueprint.local
WC Version: 8.1.1
REST API Version: ✔ 8.1.1
WC Blocks Version: ✔ 10.9.3
Action Scheduler Version: ✔ 3.6.2
Log Directory Writable: ✔
WP Version: 6.3.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.16.0
PHP Version: 8.1.23
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 8.1.2
(SecureTransport) LibreSSL/3.3.6

SUHOSIN Installed: –
MySQL Version: 8.0.16
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.1.1
WC Database Prefix: wp_
Total Database Size: 6.02MB
Database Data Size: 4.29MB
Database Index Size: 1.73MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 3.09MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.11MB + Index: 0.06MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.08MB + Index: 0.00MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 23
page: 8
post: 2
product: 18
product_variation: 7
revision: 4
shop_order: 1
shop_order_placehold: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (3) ###

WooCommerce Stripe Gateway: by WooCommerce – 7.6.0
WooCommerce: by Automattic – 8.1.1
WP Crontrol: by John Blackbourn & crontributors – 1.15.3

### Inactive Plugins (3) ###

Code Snippets: by Code Snippets Pro – 3.5.0
Order Simulator for WooCommerce: by 75nineteen Media LLC – 1.0.2
WooCommerce Blocks: by Automattic – 11.1.0

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 4.5.2
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
product-variation-management
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2023-09-24 01:29:17 +00:00
Options: ✔
Notes: 102
Onboarding: completed

### Action Scheduler ###

Complete: 21
Oldest: 2023-09-12 03:26:16 +0000
Newest: 2023-09-23 02:23:33 +0000

Pending: 2
Oldest: 2023-09-23 03:17:08 +0000
Newest: 2023-09-24 02:11:31 +0000

### Status report information ###

Generated at: 2023-09-23 02:31:17 +00:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by jorgeatorres**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds HPOS compatibility to the dashboard status widget, which hasn't been working so far unless you had compatibility mode enabled.

Closes #40382.

Screenshot 2024-02-16 at 19 19 22

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Start with a few orders in different states and different products.
2. Go to WC > Settings > Advanced > Features and make sure HPOS is enabled. If necessary, sync orders (`wp wc cot sync` via CLI) so that you can switch datastores.
3. Make sure compatibility mode is _disabled_.
4. Go to "Dashboard" on your site's admin.
5. If the "WooCommerce Status" widget isn't displayed, make sure it is enabled inside "Screen Options". You might also need to complete the onboarding steps _or_ skip them by running the following command:
   ```
   wp option update woocommerce_task_list_hidden yes
   ```
6. Make sure the widget displays correctly and totals are not `0.0`.
7. Go back to WC > Settings > Advanced > Features and switch to the posts datastore.
8. Confirm that the widget continues to display correct information.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Utilities/OrderUtil.php [hunk 1]
```diff
@@ -185,4 +184,48 @@ public static function get_table_for_orders() {
...
+
+	/**
+	 * Counts number of orders of a given type.
+	 *
+	 * @since 8.7.0
+	 *
+	 * @param string $order_type Order type.
+	 * @return array<string,int> Array of order counts indexed by order type.
+	 */
+	public static function get_count_for_type( $order_type ) {
```
Comment (vedanshujain) [inline plugins/woocommerce/src/Utilities/OrderUtil.php:196]: This method can probably use unit test.

**Review by vedanshujain — COMMENTED** [review]
Left a small suggestion, looks great otherwise!

File: plugins/woocommerce/src/Utilities/OrderUtil.php (hunk 1 above)
Comment (jorgeatorres) [reply plugins/woocommerce/src/Utilities/OrderUtil.php:196]: Unit test added in 810b427ede4fd40f203c909e81d6e0e3dca6b474.

**Comment by jorgeatorres** [comment]
Hey @vedanshujain! Thanks for the review. I've added the unit test as requested. Please to take another look when you have the chance 🐈.

**Review by vedanshujain — APPROVED** [review]

**Comment by vedanshujain** [comment]
Merging, but also flagging that we should probably evaluate these queries on the perf testing site when the build is ready!

**Pull request merged**